src/sensai/data_transformation.py

```diff
@@ -20,15 +21,75 @@ class DataFrameTransformer(ABC):
     (possibly applying the transformation to the original data frame - in-place transformation).
     A data frame transformer may require being fitted using training data.
     """
+    def __init__(self):
+        self._name = f"{self.__class__.__name__}-{id(self)}"
+        self.__isFitted = False
+        self._changesColumns = False
+        self.__removedColumns: Optional[Set[str]] = None
+        self.__addedColumns: Optional[Set[str]] = None
+
+    def getName(self) -> str:
+        """
+        :return: the name of this dft transformer, which may be a default name if the name has not been set.
+        """
+        return self._name
+
+    def setName(self, name):
+        self._name = name
 
     @abstractmethod
-    def fit(self, df: pd.DataFrame):
+    def _fit(self, df: pd.DataFrame):
         pass
 
     @abstractmethod
-    def apply(self, df: pd.DataFrame) -> pd.DataFrame:
+    def _apply(self, df: pd.DataFrame) -> pd.DataFrame:
         pass
 
+    def apply(self, df: pd.DataFrame) -> pd.DataFrame:
+        if not self.isFitted():
+            raise Exception(f"Cannot apply a DataFrameTransformer which is not fitted: "
+                            f"the df transformer {self.getName()} requires fitting")
+
+        inputColumns = set(df.columns)
+        df = self._apply(df)
+        outputColumns = set(df.columns)
+
+        if inputColumns != outputColumns:
+            self._changesColumns = True
+            self.__removedColumns = inputColumns.difference(outputColumns)
+            self.__addedColumns = outputColumns.difference(inputColumns)
+        return df
+
+    def getChangeInColumnNames(self):
+        """
+        Returns a dict describing the change in column names that was created in the most recent application of
+        the data frame transformers. If no changes in column names were created or apply was never called, returns None.
```

Comment:
It doesn't seem appropriate to return None in both of these cases. If the DFT has never been applied, an exception seems more appropriate.

Reply:
Currently I use it for the summary, which should never raise exceptions. This is also in line with a similar method in FeatureGenerator

```
    def getGeneratedColumnNames(self) -> Optional[List[str]]:
        """
        :return: Column names of the data frame generated by the most recent call of the feature generators 'generate' method.
            Returns None if generate was never called.
        """
        return self._generatedColumnNames
```
This also returns None if generate was never called

Reply:
Then you should either remove it from the summary or find another way of differentiating these cases. It's a relevant distinction in my opinion.

Reply:
It is not in line with the method you mention, because the method returns None only if it has never been called (and not in the case where nothing was generated in the most recent call).

Reply:
Right, I overlooked that. I will return an empty set in case no difference was generated

Reply:
Done